src/language-js/postprocess.js

```diff
@@ -9,12 +9,20 @@ const { composeLoc, locStart, locEnd } = require("./loc");
 const { isTypeCastComment } = require("./comments");
 
 function postprocess(ast, options) {
-  if (options.parser === "typescript" || options.parser === "flow") {
+  if (
+    options.parser === "typescript" ||
+    options.parser === "flow" ||
+    options.parser === "espree"
+  ) {
     includeShebang(ast, options);
   }
 
   // Keep Babel's non-standard ParenthesizedExpression nodes only if they have Closure-style type cast comments.
-  if (options.parser !== "typescript" && options.parser !== "flow") {
+  if (
+    options.parser !== "typescript" &&
+    options.parser !== "flow" &&
+    options.parser !== "espree"
```

Comment:
... define & use a function such as "isBabelParser" for this?

Reply:
In my opinion, this part should not used at all, relay on `parenthesized` makes [this test](https://github.com/prettier/prettier/pull/9000/files#diff-70f3836f361e3398a527b37114b863a4R54) fail on all other parsers, we need refactor it to detect from tokens, hope I can get some time to make it work.

Reply:
I am not sure how much I can understand, looking forward to the improvements (someday) thanks.

Reply:
Parser `babel` add `parenthesized` info when parsing, but other parser don't.  We use this info somewhere. Instead of using this info, we should detect if a Node is parenthesized by checking tokens around it, so we can have `parenthesized` info for all parsers. I normally use [`isParenthesized` from `eslint-utils`](https://github.com/sindresorhus/eslint-plugin-unicorn/search?q=isParenthesized&type=) for eslint parsers.